File: Priority_Queue/621. Task Scheduler/621. Task Scheduler.java
class Solution {
    // frist greedy use max heap
    public int leastInterval(char[] tasks, int n) {
        int[] chars = new int[256];
        for (char ch : tasks) {
            chars[ch]++;
        }
        PriorityQueue<Integer> pq = new PriorityQueue<>( (a, b) -> ( b- a ) );
        for (int i = 0; i < 256; i++) {
            if (chars[i] == 0) continue;
            pq.offer(chars[i]);
        }
        n++;
        
        int count = 0;
        while ( !pq.isEmpty() ) { // for every round to designate the count num of this level
            
            int loopTimes = Math.min(pq.size(), n);
            
            List<Integer> temp = new ArrayList<>();
            for (int i = 0; i < loopTimes; i++) {
                int curTask = pq.poll();
                if (curTask-1 > 0) {
                    temp.add(curTask-1);
                }
            }
            for (int i = 0; i < temp.size(); i++) pq.offer(temp.get(i)); // non-empty data continue to add into pq
            // get the task num of this round
            if (pq.size() != 0) count += n;
            else count += loopTimes;
        }
        return count;
    }
}

// n = 2
// [A B C][A B D][A B C][A D E][A B C][D E A][B C D][E F]
// A A A A A A
// B B B B B
// C C C C
// D D D D
// E E E 
// F
// ->
// A A A A A 
// B B B B 
// C C C 
// D D D D
// E E E 
// F

// A A A A A
// B B B B
// X X X X

File: Priority_Queue/621. Task Scheduler/621. Task Scheduler.cpp
class Solution {
public:
    int leastInterval(vector<char>& tasks, int n) {
        unordered_map<char,int>Map;
        for (auto ch: tasks)
            Map[ch]++;
        priority_queue<int>pq;
        for (auto a:Map) pq.push(a.second);
        
        n++;
        int count = 0;
        while (pq.size()>0)
        {                            
            int k = min(n, (int)pq.size());            
            
            vector<int>temp;
            for (int i=0; i<k; i++)
            {
                int a = pq.top();
                pq.pop();
                a--;
                if (a!=0) temp.push_back(a);
            }
            if (temp.size()>0)
                count+=n;
            else
                count+=k;
            
            for (auto x: temp)
                pq.push(x);
        }
        return count;        
//         unordered_map<char, int>Map;
//         for (auto ch : tasks) {
//             Map[ch]++;
//         }
//         priority_queue<int>pq;
//         for (auto a : Map) pq.push(a.second);
            
//         n++;
//         int count = 0;
//         while (pq.size() > 0) {
            
//             int k = min(n, (int)pq.size());
//             vector<char>temp;
//             for (int i = 0; i < k; i++) {
//                 int cur = pq.top();
//                 pq.pop();
//                 cur--;
//                 if (cur != 0) temp.push_back(cur);
//             } 
//             for (auto t : temp) pq.push(t);
//             // get the tash num of curr round
//             if (pq.size() != 0) count += n;
//             else count += k;
            
            
//         }
//         return count;
    }
};
